core/driver/sources/src/main/scala/org/virtuslab/ideprobe/ide/intellij/IntelliJProvider.scala

```diff
@@ -0,0 +1,152 @@
+  def withConfig(config: DriverConfig): IntelliJProvider
+  def withPaths(paths: IdeProbePaths): IntelliJProvider
+  def withPlugins(plugins: Plugin*): IntelliJProvider
+  def setup(): InstalledIntelliJ
+
+  // This method differentiates plugins by their root entries in zip
+  // assuming that plugins with same root entries are the same plugin
+  // and only installs last occurrance of such plugin in the list
+  // in case of duplicates.
+  protected def installPlugins(dependencies: DependencyProvider, plugins: Seq[Plugin], root: Path): Unit = {
+    val allPlugins = InternalPlugins.probePluginForIntelliJ(version) +: plugins
+
+    case class PluginArchive(plugin: Plugin, archive: Resource.Archive) {
+      val rootEntries: Set[String] = archive.rootEntries.toSet
+    }
+    val targetDir = root.resolve("plugins")
+    val archives = withParallel[Plugin, PluginArchive](allPlugins)(_.map { plugin =>
+      val file = dependencies.fetch(plugin)
+      PluginArchive(plugin, file.toArchive)
+    })
+
+    val distinctPlugins = archives.reverse.distinctBy(_.rootEntries).reverse
+
+    parallel(distinctPlugins).forEach { pluginArchive =>
+      pluginArchive.archive.extractTo(targetDir)
+      println(s"Installed ${pluginArchive.plugin}")
+    }
+  }
+
+  private def withParallel[A, B](s: Seq[A])(f: JStream[A] => JStream[B]): Seq[B] = {
+    f(parallel(s)).collect(Collectors.toList[B]).asScala.toList
+  }
+
+  private def parallel[A](s: Seq[A]) = {
+    s.asJava.parallelStream
+  }
+}
+
+final class ExistingIntelliJ(
+    dependencies: DependencyProvider,
+    path: Path,
+    override val plugins: Seq[Plugin],
+    val paths: IdeProbePaths,
+    val config: DriverConfig
+) extends IntelliJProvider {
+  override val version = IntelliJVersionResolver.version(path)
+  override def withConfig(config: DriverConfig): ExistingIntelliJ =
+    new ExistingIntelliJ(dependencies, path, plugins, paths, config)
+
+  override def withPaths(paths: IdeProbePaths): ExistingIntelliJ =
+    new ExistingIntelliJ(dependencies, path, plugins, paths, config)
+
+  override def withPlugins(plugins: Plugin*): ExistingIntelliJ =
+    new ExistingIntelliJ(dependencies, path, this.plugins ++ plugins, paths, config)
+
+  override def setup(): InstalledIntelliJ = {
+    val pluginsDir = path.resolve("plugins")
+    val backupDir = Files.createTempDirectory(path, "plugins")
+    pluginsDir.copyDir(backupDir)
+    installPlugins(dependencies, plugins, path)
+
+    new LocalIntelliJ(path, paths, config, backupDir)
+  }
+}
+
+final class IntelliJFactory(
+    dependencies: DependencyProvider,
+    override val plugins: Seq[Plugin],
+    override val version: IntelliJVersion,
+    val paths: IdeProbePaths,
+    val config: DriverConfig
+) extends IntelliJProvider {
+  override def withConfig(config: DriverConfig): IntelliJFactory =
+    new IntelliJFactory(dependencies, plugins, version, paths, config)
+
+  override def withPaths(paths: IdeProbePaths): IntelliJFactory =
```

Comment:
I wonder if adding a way to adjust `IntelliJPaths` in runtime would be a good idea for example with `withIntelliJPaths` method.

Reply:
This would be quite easy to implement for downloaded IntelliJ case, but relatively complicated for the existing.

I don't know whether adjusting paths for an existing instance makes much sense, but if it doesn't, we would have a different API for the two `IntelliJProvider` implementations, and again I don't know if the caller should be forced to know which one is being used.